### PR TITLE
LibWeb: Add preceding and following Node cases in tree constraints

### PR DESCRIPTION
**Pull request by Lubrsi**
This also does some east-const changes in TreeNode.

### CODE REVIEW

File: Userland/Libraries/LibWeb/TreeNode.h [hunk 1]
```diff
@@ -137,12 +137,30 @@ class TreeNode : public Weakable<T> {
...
+    T const* next_in_pre_order() const
     {
         return const_cast<TreeNode*>(this)->next_in_pre_order();
     }
 
-    bool is_before(const T& other) const
+    T* previous_in_pre_order()
+    {
+        T* node;
+        if ((node = previous_sibling())) {
```
Comment (alimpfard) [inline Userland/Libraries/LibWeb/TreeNode.h:148]:
```suggestion
        if (auto node = previous_sibling()) {
```
the node is only local to this `if` anyway.

File: Userland/Libraries/LibWeb/TreeNode.h [hunk 2]
```diff
@@ -153,6 +171,28 @@ class TreeNode : public Weakable<T> {
         return false;
     }
 
+    // https://dom.spec.whatwg.org/#concept-tree-preceding (Object A is 'typename U' and Object B is 'this')
+    template<typename U>
+    bool has_preceding_node_of_type_in_tree_order() const
+    {
+        for (auto* node = previous_in_pre_order(); node; node->previous_in_pre_order()) {
```
Comment (alimpfard) [inline Userland/Libraries/LibWeb/TreeNode.h:178]: This loop looks wrong
```suggestion
        for (auto* node = previous_in_pre_order(); node; node = node->previous_in_pre_order()) {
```

File: Userland/Libraries/LibWeb/TreeNode.h [hunk 3]
```diff
@@ -153,6 +171,28 @@ class TreeNode : public Weakable<T> {
...
+                return true;
+        }
+        return false;
+    }
+
+    // https://dom.spec.whatwg.org/#concept-tree-following (Object A is 'typename U' and Object B is 'this')
+    template<typename U>
+    bool has_following_node_of_type_in_tree_order() const
+    {
+        for (auto* node = next_in_pre_order(); node; node->next_in_pre_order()) {
```
Comment (alimpfard) [inline Userland/Libraries/LibWeb/TreeNode.h:189]: Same here.

File: Userland/Libraries/LibWeb/TreeNode.h (hunk 2 above)
Comment (Lubrsi) [reply Userland/Libraries/LibWeb/TreeNode.h]: Oh oops..fixed.

**Pull request merged**